File: leet-code/75. Sort Colors/SortColors75.java
package com.example;

public class SortColors75 {
    public void sortColors(int[] nums) {
        int red = 0;
        int white = 0;
        int blue = nums.length - 1;

        while (white <= blue) {
            if (nums[white] < 1) {
                swap(nums, red, white);
                red++;
                white++;
            } else if (nums[white] > 1) {
                swap(nums, white, blue);
                blue--;
            } else {
                white++;
            }
        }
    }

    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}

File: leet-code/75. Sort Colors/SortColors75.kt
package com.example.algorithm

class SortColors75 {
    fun sortColors(nums: IntArray): Unit {
        var i = 0
        var j = 0
        var k = nums.lastIndex
        while (j <= k) {
            if (nums[j] < 1) {
                swap(nums, i, j)
                i += 1
                j += 1
            } else if (nums[j] > 1) {
                swap(nums, j, k)
                k -= 1
            } else {
                j += 1
            }
        }
    }

    private fun swap(nums: IntArray, idx1: Int, idx2: Int) {
        val temp = nums[idx1]
        nums[idx1] = nums[idx2]
        nums[idx2] = temp
    }
}
